### ISSUE TITLE
Output dimension of nn.Linear doesn't stay pinned

### ISSUE THREAD

**Issue by TJHeeringa** [EIDOSLAB/simplify#7]
I am considering using your simplify package in one of my project. However, I run into an issue during a simple test of mine. I cannot get the output to stay the same size. Do you know how I need to adapt my example to get it working?

### My code: 

```python
class MyNetwork(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.first = torch.nn.Linear(5, 5)
        self.second = torch.nn.Linear(5, 5)

    def forward(self, inputs):
        return self.second(torch.relu(self.first(inputs)))

model = MyNetwork()

# set some rows of the weight matrices to zero
for m in model.modules():
    if isinstance(m, torch.nn.Linear):
        w = m.weight.data
        mask = torch.bernoulli(0.8 * torch.ones(size=(w.shape[0], 1), device=w.device))
        m.weight.data.mul_(mask)

dummy_input = torch.zeros([1,5])  # Tensor shape is that of a standard input for the given model
model(dummy_input)

to_be_modified_model = copy.deepcopy(model) # deep copy so that simplify doesn't change the original model
simplified_model = simplify(to_be_modified_model, dummy_input, pinned_out=['second', to_be_modified_module.second]) # docs are unclear about what `pinned_out` should be, so just add module name and module object
```

### Expected behavior 
Both the simplified model and the original model give the same result all the time, i.e.
```python
print(simplified_module(torch.Tensor([2,2,2,2,2])))
print(module(torch.Tensor([2,2,2,2,2])))
```
matches in dimension and size.

### Actual behavior
```python
print(simplified_module(torch.Tensor([2,2,2,2,2])))
print(module(torch.Tensor([2,2,2,2,2])))
```
gives inconsistent results. The simplified module gives the same results for all indices corresponding to nonzero rows of the second weight matrix, but the indices corresponding to the zero rows of the second weight matrix are missing. As an example: 
<img width="823" alt="image" src="https://user-images.githubusercontent.com/16029718/175975901-947672a6-8877-4bb9-bae6-b2686d02b90c.png">
for a second layer having
<img width="643" alt="image" src="https://user-images.githubusercontent.com/16029718/175976138-2c0b218e-8079-4977-b9e7-859653528f76.png">
as weight matrix and bias vector.

**Comment by TJHeeringa** [comment]
A closer look at the `__remove_zeroed_channels_hook` method seems to suggest that adding `nn.Linear` modules to `pinned_out` does nothing, and that that is the reason why I get inconsistent results. 

<img width="708" alt="image" src="https://user-images.githubusercontent.com/16029718/176134034-0d4a6862-3f85-4764-91e7-cebd652506c4.png">

**Comment by AndreaBrg** [comment]
Hi, thank you for trying our package and for the detailed issue.
What I'm failing to understand is why would you like to pin (and therefore prune) the output layer of the network.

Currently the pinning procedure was set in place to avoid problems when pruning in different manners the two branches of residual connections. Do you think it would be useful to allow for pinning other layers too?

I'm aware that the documentation is lacking and we're trying to find time to improve it, if you have suggestion we are open to them.

**Comment by TJHeeringa** [comment]
In many circumstances it is normal to want to have a small a possible output. My case deviates because I have an autoencoder. I want to remove neurons from my autoencoder and keep specific linear layers fixed so that after calling `simplify` the size of the latent space and the size of the output are still okay. In this situation being able to pin `nn.Linear` is necessary for being able to use this package.

Aside from autoencoder or some other network where you want to preserve the output size, I don't think it is really beneficial to be able to pin more layers.

On the documentation front: most of your code is self-explanatory and the main steps (fuse, propagate and remove) are clear. It would help (mostly people that try to pin `nn.Linear`) that you expand the type hint for `pinned_out` from `pinned_out: List` to either `pinned_out: List[nn.Module]` and throw a warning if you try to pin a `nn.Linear` or `pinned_out: List[Union[nn.Conv2d, nn.BatchNorm2d]]`. A little more general would be simple examples that make the workings more insightful. For example an example of a small two hidden layer neural network consisting of `nn.Conv2d` layers that gets simplified to one hidden layer. More complicated examples can be found in your paper, and don't have to be included.

**Comment by AndreaBrg** [comment]
This a use case that we did not think about, I think the it may be interesting to provide the ability to pin any kind of layer but it might be necessary to differentiate between "pin because is a residual" and "pin because I want to". I might be able to push something in the weekend/next week, or, if you want, you could do a PR.

As a side note, I was confused in the pruning of the output layer because it would completely remove a class (think of the classification layer), I guess that's not the case for you?

Thanks for the feedback on the docs.

**Comment by TJHeeringa** [comment]
I have some initial test and it seems that replacing https://github.com/EIDOSLAB/simplify/blob/3a5e9352210a79dd7074f401d98e1ae2ff71b4ff/simplify/remove.py#L22-L143
with 
```python
    # Define hook that indicates which neurons need to go
    @torch.no_grad()
    def __remove_zeroed_channels_hook(module, input_, output, name):
        input_ = input_[0][0]

        ##########################################
        #     STEP 1 - REMOVE INPUT CHANNELS     #
        ##########################################

        # Compute non-zero input channels indices
        nonzero_idx = ~(input_.view(input_.shape[0], -1).sum(dim=1) == 0)

        module.weight = torch.nn.Parameter(module.weight[:, nonzero_idx])
        module.in_features = module.weight.shape[1]

        ###########################################
        #     STEP 2 - REMOVE OUTPUT CHANNELS     #
        ###########################################

        output = torch.ones_like(output) * float('nan')

        # Compute non-zero output channels indices
        shape = module.weight.shape
        nonzero_idx = ~(module.weight.view(shape[0], -1).sum(dim=1) == 0)

        if name in pinned_outputs:
            print("Will be ignored!")
        else:
            output = torch.zeros_like(output)
            output[:, nonzero_idx] = float('nan')

            module.bias = torch.nn.Parameter(module.bias[nonzero_idx])
            module.weight = torch.nn.Parameter(module.weight[nonzero_idx])

        module.out_features = module.weight.shape[0]

        return output
```
does the trick. If I am not mistaken, this means adding a single appropriately placed if-statement is sufficient. I won't have time to do a PR in the coming month, but may have time to one later. 

This is indeed not the case for me. If you recover the MNIST written digit dataset, then it is okay to have pruned (output) layers. The edges of the images mostly have the same color anyways.

**Comment by AndreaBrg** [comment]
@TJHeeringa Hi, sorry it took so much time but I've been pretty busy at work. I just pushed to [this branch](https://github.com/EIDOSLAB/simplify/tree/extend_pinning) an update that should work for your case, I've tested it with the following code
```
import copy

import torch

from simplify import simplify

if __name__ == '__main__':
    class MyNetwork(torch.nn.Module):
        def __init__(self):
            super().__init__()
            self.first = torch.nn.Linear(5, 5)
            self.second = torch.nn.Linear(5, 5)

        def forward(self, inputs):
            return self.second(torch.relu(self.first(inputs)))


    model = MyNetwork()

    # set some rows of the weight matrices to zero
    for m in model.modules():
        if isinstance(m, torch.nn.Linear):
            w = m.weight.data
            mask = torch.bernoulli(0.8 * torch.ones(size=(w.shape[0], 1), device=w.device))
            m.weight.data.mul_(mask)

    dummy_input = torch.zeros([1, 5])  # Tensor shape is that of a standard input for the given model
    model(dummy_input)

    to_be_modified_model = copy.deepcopy(model)  # deep copy so that simplify doesn't change the original model
    simplified_model = simplify(to_be_modified_model, dummy_input, pinned_out=["second"])

    print(simplified_model(torch.Tensor([2, 2, 2, 2, 2])))
    print(model(torch.Tensor([2, 2, 2, 2, 2])))
```
I've tried to keep the logic we used for pinning conv and bn layers i.e. the operations are performed with smaller matrices but the output contains zeroes where the neurons were pruned.

Let me know if it works.

**Comment by TJHeeringa** [comment]
I did several runs with the code you tested it with. I noticed that
 1. If `model.first.weight` has no zero rows, then the model and simplified model agree regardless of whether `model.second.weight`  has zero rows.
 2. If `model.first.weight` has zero rows, then the model and simplified model disagree on all elements corresponding to the non zero rows of `model.second.weight` and agree on the elements corresponding to the zero rows of `model.second.weight`. 

Hence, in the second case the simplification does not work. My guess is that this has to do with the biases of the first layer not being propagated properly. 

To test this guess I made the weight matrices 2x2 instead of 5x5, and ran the code until I got a zero matrix for `module.first.weight` and had one zero and one nonzero row for `module.second.weight`. This is a situation that the model and simplified model disagree:
```python
tensor([-0.1632, -0.1126], grad_fn=<AddBackward0>)       # simplified model
tensor([-0.1632, -0.5466], grad_fn=<AddBackward0>)       # model
```
The weights and biases of the model and the simplified model are given by
```python
 --- ORIGINAL MODEL PARAMETERS --- 
Parameter containing:
tensor([[0., -0.],
        [-0., -0.]], requires_grad=True)
Parameter containing:
tensor([0.4721, 0.4603], requires_grad=True)
Parameter containing:
tensor([[-0.0000, -0.0000],
        [-0.4141, -0.5182]], requires_grad=True)
Parameter containing:
tensor([-0.1632, -0.1126], requires_grad=True)
 
 --- SIMPLIFIED MODEL PARAMETERS --- 
Parameter containing:
tensor([], size=(0, 2), requires_grad=True)
Parameter containing:
tensor([], requires_grad=True)
Parameter containing:
tensor([], size=(0, 0), requires_grad=True)
Parameter containing:
tensor([], requires_grad=True)
```
This shows that the simplified model just returns `module.second.bias`, whereas the model computes `module.second.weight * module.first.bias + module.second.bias`. Thus, the bias of pruned layers seems to be removed too in the simplified models. This seems to me as the cause of the disagreement between the models and the simplified models.

I am not sure why running 
```python
    print(simplified_model.first.weight)
    print(simplified_model.first.bias)
    print(simplified_model.second.weight)
    print(simplified_model.second.bias)
```
gives
```python
 --- SIMPLIFIED MODEL PARAMETERS --- 
Parameter containing:
tensor([], size=(0, 2), requires_grad=True)
Parameter containing:
tensor([], requires_grad=True)
Parameter containing:
tensor([], size=(0, 0), requires_grad=True)
Parameter containing:
tensor([], requires_grad=True)
```
I would have expected to at least get `tensor([-0.1632, -0.1126], requires_grad=True)` in the last line. Does this have to do with the biases being improperly propagated/removed?

**Comment by TJHeeringa** [comment]
The code is used for the previous comment is:
```python
import copy

import torch

from simplify import simplify

SIZE = 2
if __name__ == '__main__':
    class MyNetwork(torch.nn.Module):
        def __init__(self):
            super().__init__()
            self.first = torch.nn.Linear(SIZE, SIZE)
            self.second = torch.nn.Linear(SIZE, SIZE)

        def forward(self, inputs):
            return self.second(torch.relu(self.first(inputs)))


    model = MyNetwork()

    # set some rows of the weight matrices to zero
    for m in model.modules():
        if isinstance(m, torch.nn.Linear):
            w = m.weight.data
            mask = torch.bernoulli((1-1/SIZE) * torch.ones(size=(w.shape[0], 1), device=w.device))
            m.weight.data.mul_(mask)

    dummy_input = torch.zeros([1, SIZE])  # Tensor shape is that of a standard input for the given model
    model(dummy_input)

    to_be_modified_model = copy.deepcopy(model)  # deep copy so that simplify doesn't change the original model
    simplified_model = simplify(to_be_modified_model, dummy_input, pinned_out=["second"])

    print(" --- SIMPLIFICATION TEST RESULTS --- 2")
    print(simplified_model(torch.Tensor([2]*SIZE)))
    print(model(torch.Tensor([2]*SIZE)))
    print(" ")

    print(" --- SIMPLIFICATION TEST RESULTS --- 1")
    print(simplified_model(torch.Tensor([1]*SIZE)))
    print(model(torch.Tensor([1]*SIZE)))
    print(" ")

    print(" --- ORIGINAL MODEL PARAMETERS --- ")
    print(model.first.weight)
    print(model.first.bias)
    print(model.second.weight)
    print(model.second.bias)
    print(" ")

    print(" --- SIMPLIFIED MODEL PARAMETERS --- ")
    print(simplified_model.first.weight)
    print(simplified_model.first.bias)
    print(simplified_model.second.weight)
    print(simplified_model.second.bias)
    print(" ")
```

**Comment by AndreaBrg** [comment]
@TJHeeringa sorry for the very late reply but between NIPS and vacations I had no available time :(
Thank you for the code, I'll check it out and I'll try to find out what's wrong.

**Comment by AndreaBrg** [comment]
@TJHeeringa I found out a very weird bug which should be fixed in https://github.com/EIDOSLAB/simplify/commit/bf8aa4cfdf0780d6653bf2c0ad45cc177c61b11b

With your code I get the following output (I added some prints, but the rest is mostly unchanged):
 ```
--- SIMPLIFICATION TEST RESULTS --- 2
Dense
tensor([-0.1141, -0.1334], grad_fn=<AddBackward0>)
Simplified
tensor([-0.1141, -0.1334], grad_fn=<AddBackward0>)
 
 --- SIMPLIFICATION TEST RESULTS --- 1
Dense
tensor([-0.1088, -0.1099], grad_fn=<AddBackward0>)
Simplified
tensor([-0.1088, -0.1099], grad_fn=<AddBackward0>)
 
 --- ORIGINAL MODEL PARAMETERS --- 
first
Parameter containing:
tensor([[-0.0053,  0.3793],
        [-0.0000, -0.0000]], requires_grad=True)
Parameter containing:
tensor([-0.2723,  0.1896], requires_grad=True)
 
second
Parameter containing:
tensor([[-0.0140,  0.5607],
        [-0.0628,  0.1871]], requires_grad=True)
Parameter containing:
tensor([-0.2137, -0.1390], requires_grad=True)
 
 --- SIMPLIFIED MODEL PARAMETERS --- 
first
Parameter containing:
tensor([[-0.0053,  0.3793]], requires_grad=True)
Parameter containing:
tensor([-0.2723], requires_grad=True)
 
second
Parameter containing:
tensor([[-0.0140],
        [-0.0628]], requires_grad=True)
Parameter containing:
tensor([-0.1074, -0.1035], requires_grad=True)
```
and I pruned the model using this masks:
```
if n == "first":
    mask = torch.tensor([
        [1, 1],
        [0, 0]
    ])
if n == "second":
    mask = torch.tensor([
        [1, 1],
        [1, 1]
    ])
```

**Comment by TJHeeringa** [comment]
I have generated 10k fully connected networks with width up to 10 and up to 10 layers, and pruned them with random masks. With this fix it works for all these cases. However, in 0.3% of cases some entries in the output differed by about 1e-8. To me that seems as a numerical error originating from propagating the biases through the network. That is good enough for me. So, I think the issue is resolved.

**Comment by AndreaBrg** [comment]
Yes, while experimenting we also found an error of that magnitude between the original and the simplified network and concluded that numerical approximations are the cause. We usually test with `torch.allclose` using a tolerance between 1e-7 and 1e-6.

Glad that it now works!
